contribute/engineering/backend/instrumentation.md

```diff
@@ -0,0 +1,207 @@
+
+Use the _pkg/infra/log_ package to create a named structured logger. Example:
+
+```go
+import (
+  "fmt"
+
+  "github.com/grafana/grafana/pkg/infra/log"
+)
+
+logger := log.New("my-logger")
+logger.Debug("Debug msg")
+logger.Warning("Warning msg")
+logger.Info("Info msg")
+logger.Error("Error msg", "error", fmt.Errorf("BOOM))
+```
+
+### Naming conventions
+
+Consider naming the logger using lowercase characters, e.g. `log.New("my-logger")` using snake_case or kebab-case styling.
+
+Consider to prefix the logger name with an area name if using different loggers across a feature/common packages, e.g. `log.New("plugin.loader")` and `log.New("plugin.client")`.
+
+Start the log message with a capital letter, e.g. `logger.Info("Hello world")` instead of `logger.Info("hello world")`.
+
+Use snake_case style when naming log keys, e.g. _remote_addr_ instead of _remoteAddr_.
+
+Use the key _error_ when logging Go errors, e.g. `logger.Error("Something failed", "error", fmt.Errorf("BOOM))`.
+
+### Log levels
+
+When to use which log level?
+
+- **Debug:** Informational messages of high frequency and/or less-important messages during normal operations.
+- **Info:** Informational messages of low frequency and/or important messages.
+- **Error:** Error messages indicating some operation failed (with an error).
+- **Warning:** Should in normal cases not be used/needed. If used should be actionable.
+
+### Contextual logging
+
+Use a contextual logger to include additional key/value pairs attached to `context.Context`, e.g. `traceID`, to allow correlating logs with traces and/or correlate logs with a common identifier.
+
+Example:
+
+```go
+import (
+  "context"
+  "fmt"
+
+  "github.com/grafana/grafana/pkg/infra/log"
+)
+
+var logger = log.New("my-logger")
+
+func doSomething(ctx context.Context) {
+  ctxLogger := logger.FromContext(ctx)
+  ctxLogger.Debug("Debug msg")
+  ctxLogger.Warning("Warning msg")
+  ctxLogger.Info("Info msg")
+  ctxLogger.Error("Error msg", "error", fmt.Errorf("BOOM))
+}
+```
+
+### Enable certain log levels for certain loggers
+
+During development it's convenient to enable certain log level, e.g. debug, for certain loggers to minimize the generated log output and make it easier to find things. See [[log.filters]](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#filters) for information how to configure this.
+
+It's also possible to configure multiple loggers:
+
+```ini
+[log]
+filters = rendering:debug \
+          ; alerting.notifier:debug \
+          oauth.generic_oauth:debug \
+          ; oauth.okta:debug \
+          ; tsdb.postgres:debug \
+          ; tsdb.mssql:debug \
+          ; provisioning.plugins:debug \
+          ; provisioning:debug \
+          ; provisioning.dashboard:debug \
+          ; provisioning.datasources:debug \
+          datasources:debug \
+          data-proxy-log:debug
+```
+
+## Metrics
```

Comment:
Given the existence of https://grafana.com/docs/loki/latest/logql/metric_queries/ and https://grafana.com/docs/grafana/latest/enterprise/recorded-queries/, I don't have a clear picture on why would I ever publish metrics, as I can just log everything and then create metrics from the logs as needed. Could we maybe add a section explaining when to use one over the other?

Reply:
Creating metrics from logs can be great but in some cases metrics are just better. Ex
* When instrumenting something very high frequent as it would spam the logs. 
* Gauges. Ex how many requests there are in flight or how many connections are currently waiting for the database connection pool. Doing this with logs would require pre and post logs and calculating the diff. 


Another thing to consider are none grafana cloud users of Grafana (I know!!) who might not have loki with recording rules.